File: Longest_Substring_with_at_most_K_Distinct_Characters.cpp
/*340. Longest Substring with At Most K Distinct Characters  QuestionEditorial Solution  My Submissions
Total Accepted: 8198 Total Submissions: 20927 Difficulty: Hard
Given a string, find the length of the longest substring T that contains at most k distinct characters.

For example, Given s = “eceba” and k = 2,

T is "ece" which its length is 3.

Hide Company Tags Google
Hide Tags Hash Table String
Hide Similar Problems (H) Longest Substring with At Most Two Distinct Characters
*/



/*use array/hashmap to store each character value
array is more quickly
array --- char is 1 byte = 8 bits = 256 possible values
once this character is not exist before, ++count
once the number is larger than k, delete from the start of string
check the maxLen of the substring for each step*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        vector<int> ss(256);
        int count = 0;
        int i = 0;
        int ret = 0;
        for(int j = 0; j < s.size(); ++j) {
            if(ss[s[j]]++ == 0) count++;
            if(count > k){
                while(--ss[s[i++]] > 0);
                count--;
            }
            ret = max(ret, j-i+1);
        }
        return ret;
    }
};




/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        //array
        int[] ss = new int[256];
        int count = 0;
        int i = 0;
        int ret = 0;
        for(int j = 0; j < s.length(); ++j){
            if(ss[s.charAt(j)]++ == 0) count++;

            if(count >k){
                while(--ss[s.charAt(i++)] > 0);
                count--;
            }
            /*while(count > k){
                char c1 = s.charAt(index);
                if(--str[c1] == 0){
                    count--;
                }
                index++;
            }*/
            ret = Math.max(ret, j-i+1);
        }
        return ret;
    }
}




public class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        //hashmap
        Map<Character, Integer> mymap = new HashMap<>();
        int maxLen = 0;
        int index = 0;
        
        for(int i = 0; i < s.length(); ++i){
            char c = s.charAt(i);
            if(!mymap.containsKey(c)){
                mymap.put(c, 0);
            }
            mymap.put(c, mymap.get(c)+1);
            
            while(mymap.size()>k){
                char c1 = s.charAt(index);
                mymap.put(c1, mymap.get(c1)-1);
                if(mymap.get(c1)==0){
                    mymap.remove(c1);
                }
                index++;
            }
            maxLen = Math.max(maxLen, i-index+1);
        }
        return maxLen;
    }
}

File: Longest_Substring_with_At_Most_K_Distinct_Characters.swift
// Given a string, find the length of the longest substring T that contains at most k distinct characters.

// Example 1:

// Input: s = "eceba", k = 2
// Output: 3
// Explanation: T is "ece" which its length is 3.
// Example 2:

// Input: s = "aa", k = 1
// Output: 2
// Explanation: T is "aa" which its length is 2.

// Solution1 : 2 pointers with map
// 
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
    func lengthOfLongestSubstringKDistinct(_ s: String, _ k: Int) -> Int {
        guard !s.isEmpty, k > 0 else { return 0 }
        
        var s = Array(s)
        var map = [Character: Int]()
        var left = 0
        var right = 0
        var dis = -1
        
        while right < s.count {
            map[s[right], default: 0] += 1
            
            while left <= right, map.keys.count > k {
                map[s[left]]! -= 1
                if map[s[left]]! == 0 {
                    map[s[left]] = nil
                }
                left += 1
            }
            dis = max(dis, right - left + 1)
            right += 1
        }
        return dis == -1 ? 0 : dis
    }
}